core/embed/rust/src/ui/shape/bitmap/bitmap_base.rs

```diff
@@ -0,0 +1,313 @@
+use crate::trezorhal::bitblt::BitBlt;
+
+use crate::ui::{display::Color, geometry::Offset};
+
+use core::{cell::Cell, marker::PhantomData};
+
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub enum BitmapFormat {
+    /// 1-bit mono
+    MONO1,
+    /// 1-bit mono packed (bitmap stride is in bits)
+    MONO1P,
+    /// 4-bit mono
+    MONO4,
+    /// 8-bit mono
+    MONO8,
+    /// 16-bit color, RGB565 format
+    RGB565,
+    /// 32-bit color, RGBA format
+    RGBA8888,
+}
+
+pub struct Bitmap<'a> {
+    /// Pointer to top-left pixel
+    ptr: *mut u8,
+    /// Stride in bytes
+    stride: usize,
+    /// Size in pixels
+    size: Offset,
+    /// Format of pixels
+    format: BitmapFormat,
+    /// Bitmap data is mutable
+    mutable: bool,
+    /// DMA operation is pending
+    dma_pending: Cell<bool>,
+    ///
+    _phantom: core::marker::PhantomData<&'a ()>,
+}
+
+impl<'a> Bitmap<'a> {
+    /// Creates a new bitmap referencing a specified buffer.
+    ///
+    /// Optionally minimal height can be specified and then the height
+    /// of the new bitmap is adjusted to the buffer size.
+    ///
+    /// Returns None if the buffer is not big enough.
+    ///
+    /// The `buff` needs to be properly aligned and big enough
+    /// to hold a bitmap with the specified format and size
+    pub fn new(
+        format: BitmapFormat,
+        stride: Option<usize>,
+        mut size: Offset,
+        min_height: Option<i16>,
+        buff: &'a [u8],
+    ) -> Option<Self> {
+        if size.x < 0 && size.y < 0 {
+            return None;
+        }
+
+        let min_stride = match format {
+            BitmapFormat::MONO1 => (size.x + 7) / 8,
+            BitmapFormat::MONO1P => 0,
+            BitmapFormat::MONO4 => (size.x + 1) / 2,
+            BitmapFormat::MONO8 => size.x,
+            BitmapFormat::RGB565 => size.x * 2,
+            BitmapFormat::RGBA8888 => size.x * 4,
+        } as usize;
+
+        let stride = stride.unwrap_or(min_stride);
+
+        let alignment = match format {
+            BitmapFormat::MONO1 => 1,
+            BitmapFormat::MONO1P => 1,
+            BitmapFormat::MONO4 => 1,
+            BitmapFormat::MONO8 => 1,
+            BitmapFormat::RGB565 => 2,
+            BitmapFormat::RGBA8888 => 4,
+        };
+
+        assert!(stride >= min_stride);
+        assert!(buff.as_ptr() as usize & (alignment - 1) == 0);
+        assert!(stride & (alignment - 1) == 0);
+
+        let max_height = if stride == 0 {
+            size.y as usize
+        } else {
+            buff.len() / stride
+        };
+
+        if size.y as usize > max_height {
+            if let Some(min_height) = min_height {
+                if max_height >= min_height as usize {
+                    size.y = max_height as i16;
+                } else {
+                    return None;
+                }
+            } else {
+                return None;
+            }
+        }
```

Comment:
/// Optionally minimal height can be specified and then the height
 /// of the new bitmap is adjusted to the buffer size.

In several places in the code, we have a preallocated buffer (at the bump) and want to create a bitmap from it. We aim for a bitmap of specified width - the height is less important. We just need a minimum number of rows (for example - 1) and do not need to allocate more than a certain number of rows.

Perhaps it would be better to have two functions, but I think that would complicate the code elsewhere.

Reply:
In general, we use this functionality where the bitmap serves as a temporary buffer, either for decompression or in a progressive renderer. We need a bitmap with at least one row, but we don't need a bitmap with more rows than the height of the original bitmap. Thus, we request a bitmap with a height ranging from 1 to something, and `new()` returns what we can afford according to the size of the provided buffer.

Reply:
so why not explicitly request a bitmap with just one row?

Reply:
In such situations: I don't want to just one row - it's just minimum I need. I want as much as possible, usually to speed the operation. But need not to get more than something, because it's useless.